(DOCSP-40807) Indexes > Compound indexes (#63)
mongodb/docs-cpp: source/indexes/compound-index.txt

.. _cpp-compound-index:

================
Compound Indexes
================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: index, query, optimization, efficiency

Overview
--------

**Compound indexes** are indexes that hold references to multiple
fields within a collection's documents. These indexes improve multi-field query and
sort performance.

> To create a compound index, call the ``create_index()`` method and specify a document containing: 

Reviewer: S: We try to use full sentences when [introducing lists](https://www.mongodb.com/docs/meta/style-guide/style/lists/#introductory-text), so you can reword to:
```suggestion
To create a compound index, call the ``create_index()`` method and specify a document containing the following information: 
```

> - The fields on which to create the index
> - The sort order for the indexed values, ``1`` for ascending or ``-1`` for descending.

Reviewer: S: we also try to avoid [starting lists with articles](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items)
```suggestion
- Fields on which to create the index
- Sort order for the indexed values, ``1`` for ascending or ``-1`` for descending.
```

> - Fields on which to create the index
> - Sort order for the indexed values, ``1`` for ascending or ``-1`` for descending.

Reviewer: S: I'd separate the second bullet into 2 sentences. Also, since the second one has periods, [add a period](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items) to the end of the first bullet point
```suggestion
- Fields on which to create the index.
- Sort order for the indexed values. Use ``1`` for ascending or ``-1`` for descending.
```

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``movies`` collection in the ``sample_mflix``
database from the :atlas:`Atlas sample datasets </sample-data>`. To access this collection
from your C++ application, instantiate a ``mongocxx::client`` that connects to an Atlas cluster
and assign the following values to your ``db`` and ``collection`` variables:

.. literalinclude:: /includes/indexes/indexes.cpp
    :start-after: start-db-coll
    :end-before: end-db-coll
    :language: cpp
    :copyable:
    :dedent:

To learn how to create a free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

Create Compound Index
---------------------

The following example creates an ascending compound index on the ``title`` and ``year`` fields:

.. literalinclude:: /includes/indexes/indexes.cpp
   :start-after: start-index-compound
   :end-before: end-index-compound
   :language: cpp
   :copyable:
   :dedent:

> The following is an example of a query that is covered by the index
> created in the preceding code example:

Reviewer: S: I'd reword this slightly to remove some repetition
```suggestion
The following query is covered by the index
created in the preceding code example:
```

.. io-code-block::
   :copyable: true

   .. input:: /includes/indexes/indexes.cpp
      :start-after: start-index-compound-query
      :end-before: end-index-compound-query
      :language: php
      :dedent:
   
   .. output::
      :language: cli
      :visible: false

      { "_id" :..., "plot" : "Peter Pan enters the nursery of the Darling children...", 
      ..., "year" : 1924, "imdb" : ..., "type", "movie",...}
      
Additional Information
----------------------

> To learn more about compound indexes, see :manual:`Compound Indexes </core/index-compound>` in the {+mdb-server+} manual.
> 
> To view runnable examples that demonstrate how to manage indexes, see
> :ref:`cpp-indexes`.
> 
> To learn more about indexes in MongoDB, see the :manual:`Indexes </indexes>`
> guide in the {+mdb-server+} manual.

Reviewer: S: you can group the Server links together. Maybe reorder like:
```suggestion
To view runnable examples that demonstrate how to manage indexes, see 
:ref:`cpp-indexes`.

To learn more about indexes, see the following resources in the {+mdb-server+} manual:

- :manual:`Indexes </indexes>`
- :manual:`Compound Indexes </core/index-compound>`
```
